File: Java/Ema/PerfTools/src/main/java/com/refinitiv/ema/perftools/common/DirectoryHandler.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.md for details.                  --
 *|           Copyright (C) 2022 Refinitiv. All rights reserved.         	  --
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.ema.perftools.common;

import com.refinitiv.ema.access.AckMsg;
import com.refinitiv.ema.access.Data;
import com.refinitiv.ema.access.ElementEntry;
import com.refinitiv.ema.access.ElementList;
import com.refinitiv.ema.access.EmaFactory;
import com.refinitiv.ema.access.FilterEntry;
import com.refinitiv.ema.access.FilterList;
import com.refinitiv.ema.access.GenericMsg;
import com.refinitiv.ema.access.Map;
import com.refinitiv.ema.access.MapEntry;
import com.refinitiv.ema.access.OmmConsumerClient;
import com.refinitiv.ema.access.OmmConsumerEvent;
import com.refinitiv.ema.access.OmmState;
import com.refinitiv.ema.access.RefreshMsg;
import com.refinitiv.ema.access.ReqMsg;
import com.refinitiv.ema.access.StatusMsg;
import com.refinitiv.ema.access.UpdateMsg;
import com.refinitiv.ema.access.DataType.DataTypes;
import com.refinitiv.ema.rdm.EmaRdm;

/**
 * This is the source directory handler for the EMA consumer application. It
 * provides methods for sending the source directory request to a provider and
 * processing the response. Methods for setting the service name, getting the
 * service information, and closing a source directory stream are also provided.
 */
public class DirectoryHandler implements OmmConsumerClient
{
    private static final long FILTER_TO_REQUEST = EmaRdm.SERVICE_INFO_FILTER  |
    		EmaRdm.SERVICE_STATE_FILTER | EmaRdm.SERVICE_GROUP_FILTER;
    
    private String _serviceName ;    // requested service
    private int _serviceId;	 //requested service id
    class ServiceState
    {
    	int streamState;
    	int dataState;
    	int statusCode;
    	String text;
    }
    private ServiceState _state;            // requested service state
    private ReqMsg _directoryRequest = EmaFactory.createReqMsg();
    private boolean _acceptRequest;
    
    /**
     * Instantiates a new directory handler.
     */
    public DirectoryHandler()
    {
        _state = new ServiceState();
    }

    /**
     * Sets the service name requested by the application.
     * 
     * @param servicename - The service name requested by the application
     */
    public void serviceName(String servicename)
    {
        _serviceName = servicename;
    }

    /**
     * Checks if is requested service up.
     *
     * @return true if service requested by application is up, false if not.
     */
    public boolean isRequestedServiceUp()
    {
        return  _acceptRequest && _state.streamState == OmmState.StreamState.OPEN && _state.dataState == OmmState.DataState.OK;
    }

    /**
     * Service id.
     *
     * @return service id associated with the service name requested by application.
     */
    public int serviceId()
    {
        return _serviceId;
    }

    /**
     * Sends a source directory request to a channel. This consists of getting a
     * message buffer, encoding the source directory request, and sending the
     * source directory request to the server.
     *
     * @return the request
     */
    public ReqMsg getRequest()
    {
        //initialize directory _state
        //this will be updated as refresh and status messages are received
        _state.dataState = OmmState.DataState.NO_CHANGE;
        _state.streamState = OmmState.StreamState.CLOSED;
        
        _directoryRequest.clear();
        _directoryRequest.domainType(EmaRdm.MMT_DIRECTORY).serviceName(_serviceName).filter(FILTER_TO_REQUEST).interestAfterRefresh(true);

        return _directoryRequest;
    }
    
	private void decode(Map map)
	{
		FilterList filterList;
		boolean findMapEntry = false;
		for(MapEntry mapEntry : map)
		{
				if (mapEntry.loadType() == DataTypes.FILTER_LIST)
				{
					filterList = mapEntry.filterList();
					for(FilterEntry filterEntry : filterList)
					{
							if (filterEntry.loadType() == DataTypes.ELEMENT_LIST )
							{
								if (decode(filterEntry.elementList()))
								{
									_serviceId = (int) mapEntry.key().uintValue();
									findMapEntry = true;
								}
							}
					}
				}
				
				if (findMapEntry)
					return;
		}
	}

	private boolean decode(ElementList elementList)
	{
		boolean foundService = false;
		for (ElementEntry elementEntry : elementList)
		{
			if (Data.DataCode.BLANK != elementEntry.code())
			{
				switch (elementEntry.loadType())
				{
					case DataTypes.ASCII :
						if (elementEntry.name().equals("Name") && elementEntry.ascii().ascii().equals(_serviceName))
							foundService = true;
						break;
					case DataTypes.UINT :
						if (elementEntry.name().equals("AcceptingRequests"))
						{
							if (elementEntry.uintValue() == 1)
								_acceptRequest = true;
							else
								_acceptRequest = false;
						}
						break;
					default :
						break;
				}
			}
		}
		return foundService;
	}

	/* (non-Javadoc)
	 * @see com.refinitiv.ema.access.OmmConsumerClient#onRefreshMsg(com.refinitiv.ema.access.RefreshMsg, com.refinitiv.ema.access.OmmConsumerEvent)
	 */
	@Override
	public void onRefreshMsg(RefreshMsg refreshMsg, OmmConsumerEvent consumerEvent)
	{
		 System.out.println("Received Source Directory Refresh\n");
		 
		 _state.dataState = refreshMsg.state().dataState();
	     _state.streamState = refreshMsg.state().streamState();
	         
	     if (refreshMsg.payload().dataType() != DataTypes.MAP)
	        	System.out.println("Received invalid source directory refresh msg.");
	     else
	        	decode(refreshMsg.payload().map());
         
         if (!isRequestedServiceUp())
             System.out.println("Requested service '" + _serviceName + "' not up. Waiting for service to be up...");
	}

	/* (non-Javadoc)
	 * @see com.refinitiv.ema.access.OmmConsumerClient#onUpdateMsg(com.refinitiv.ema.access.UpdateMsg, com.refinitiv.ema.access.OmmConsumerEvent)
	 */
	@Override
	public void onUpdateMsg(UpdateMsg updateMsg, OmmConsumerEvent consumerEvent)
	{
		System.out.println("Received Source Directory Update\n");

        if (updateMsg.payload().dataType() != DataTypes.MAP)
        	System.out.println("Received invalid source directory update msg.");
        else
        	decode(updateMsg.payload().map());
	}

	/* (non-Javadoc)
	 * @see com.refinitiv.ema.access.OmmConsumerClient#onStatusMsg(com.refinitiv.ema.access.StatusMsg, com.refinitiv.ema.access.OmmConsumerEvent)
	 */
	@Override
	public void onStatusMsg(StatusMsg statusMsg, OmmConsumerEvent consumerEvent)
	{
		 System.out.println("Received Source Directory Status\n");
         
		 System.out.println(statusMsg.toString());
		 
         if (statusMsg.hasState())
         {
             _state.dataState = statusMsg.state().dataState();
             _state.streamState = statusMsg.state().streamState();
         }
	}

	/* (non-Javadoc)
	 * @see com.refinitiv.ema.access.OmmConsumerClient#onGenericMsg(com.refinitiv.ema.access.GenericMsg, com.refinitiv.ema.access.OmmConsumerEvent)
	 */
	@Override
	public void onGenericMsg(GenericMsg genericMsg, OmmConsumerEvent consumerEvent) {}
	
	/* (non-Javadoc)
	 * @see com.refinitiv.ema.access.OmmConsumerClient#onAckMsg(com.refinitiv.ema.access.AckMsg, com.refinitiv.ema.access.OmmConsumerEvent)
	 */
	@Override
	public void onAckMsg(AckMsg ackMsg, OmmConsumerEvent consumerEvent) {}
	
	/* (non-Javadoc)
	 * @see com.refinitiv.ema.access.OmmConsumerClient#onAllMsg(com.refinitiv.ema.access.Msg, com.refinitiv.ema.access.OmmConsumerEvent)
	 */
	@Override
	public void onAllMsg(com.refinitiv.ema.access.Msg msg, OmmConsumerEvent consumerEvent) {}
}

File: CSharp/Ema/Src/Core/Access/DirectoryHandler.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2024-2025 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

using LSEG.Eta.Codec;
using LSEG.Eta.Common;
using LSEG.Eta.ValueAdd.Rdm;
using LSEG.Eta.ValueAdd.Reactor;
using System.Collections.Generic;
using System.Text;
using Buffer = LSEG.Eta.Codec.Buffer;

namespace LSEG.Ema.Access
{
    internal class DirectoryHandler : IDirectoryMsgCallback
    {
        private static string CLIENT_NAME = "DirectoryHandler";

        internal OmmServerBaseImpl m_ServerBaseImpl;

        private DirectoryStatus m_directoryStatus =  new DirectoryStatus();
        private DirectoryRefresh m_directoryRefresh = new DirectoryRefresh();
        private ReactorSubmitOptions m_rsslSubmitOptions = new ReactorSubmitOptions(); 
        private Buffer m_statusText = new Buffer();

        private bool m_apiAdminControl;

        internal List<ItemInfo> ItemInfoList = new List<ItemInfo>();

        public DirectoryHandler(OmmServerBaseImpl serverBaseImpl)
        {
            m_ServerBaseImpl = serverBaseImpl;
            m_statusText.Data("Source Directory Refresh Completed");
        }

        public void Initialize()
        {
            m_apiAdminControl = m_ServerBaseImpl.ConfigImpl.AdminControlDirectory == OmmIProviderConfig.AdminControlMode.API_CONTROL;
        }

        public ReactorCallbackReturnCode RdmDirectoryMsgCallback(RDMDirectoryMsgEvent directoryMsgEvent)
        {
            m_ServerBaseImpl.EventReceived();
            using var lockScope = m_ServerBaseImpl.GetUserLocker().EnterLockScope();
            StringBuilder temp = m_ServerBaseImpl.GetStrBuilder();

            ItemInfo? itemInfo = null;

            ClientSession clientSession = (ClientSession)directoryMsgEvent.ReactorChannel!.UserSpecObj!;
            DirectoryMsg? directoryMsg = directoryMsgEvent.DirectoryMsg;

            if (directoryMsg == null || directoryMsgEvent.Msg == null)
            {
                temp.Append("Directory message rejected - Invalid directory domain message.");

                if (directoryMsgEvent.Msg != null)
                {
                    SendDirectoryReject(directoryMsgEvent.ReactorChannel, directoryMsgEvent.Msg.StreamId, StateCodes.USAGE_ERROR, temp.ToString());
                }

                if (m_ServerBaseImpl.GetLoggerClient().IsErrorEnabled)
                {
                    temp.Append(ILoggerClient.CR).Append("Stream Id ").Append(directoryMsgEvent.Msg?.StreamId)
                    .Append(ILoggerClient.CR).Append("Client handle ").Append(clientSession.ClientHandle)
                    .Append(ILoggerClient.CR).Append("Instance Name ").Append(m_ServerBaseImpl.InstanceName);

                    m_ServerBaseImpl.GetLoggerClient().Trace(CLIENT_NAME, temp.ToString());
                }

                return ReactorCallbackReturnCode.SUCCESS;
            }

            if (!m_ServerBaseImpl.ConfigImpl.IProviderConfig.AcceptMessageWithoutBeingLogin && !clientSession.IsLogin)
            {
                temp.Append("Directory message rejected - there is no logged in user for this session.");

                SendDirectoryReject(directoryMsgEvent.ReactorChannel, directoryMsgEvent.Msg.StreamId, StateCodes.USAGE_ERROR,
                        temp.ToString());

                if (m_ServerBaseImpl.GetLoggerClient().IsTraceEnabled)
                {
                    temp.Append(ILoggerClient.CR).Append("Stream Id ").Append(directoryMsgEvent.Msg.StreamId)
                    .Append(ILoggerClient.CR).Append("Client handle ").Append(clientSession.ClientHandle)
                    .Append(ILoggerClient.CR).Append("Instance Name ").Append(m_ServerBaseImpl.InstanceName);

                    m_ServerBaseImpl.GetLoggerClient().Trace(CLIENT_NAME,
                            temp.ToString());
                }

                return ReactorCallbackReturnCode.SUCCESS;
            }

            switch (directoryMsg.DirectoryMsgType)
            {
                case DirectoryMsgType.REQUEST:
                    {
                        if (m_ServerBaseImpl.GetLoggerClient().IsTraceEnabled)
                        {
                            temp.Append("Received directory request message.")
                                .Append(ILoggerClient.CR).Append("Stream Id ").Append(directoryMsgEvent.Msg.StreamId)
                                .Append(ILoggerClient.CR).Append("Client handle ").Append(clientSession.ClientHandle)
                                .Append(ILoggerClient.CR).Append("Instance Name ").Append(m_ServerBaseImpl.InstanceName);

                            m_ServerBaseImpl.GetLoggerClient().Trace(CLIENT_NAME, temp.ToString());
                        }

                        itemInfo = clientSession.GetItemInfo(directoryMsgEvent.Msg.StreamId);
                        DirectoryRequest directoryRequest = directoryMsg.DirectoryRequest!;

                        if (m_ServerBaseImpl.ConfigImpl.IProviderConfig.AcceptDirMessageWithoutMinFilters ||
                                ((directoryRequest.Filter & Eta.Rdm.Directory.ServiceFilterFlags.INFO) == Eta.Rdm.Directory.ServiceFilterFlags.INFO)
                                && ((directoryRequest.Filter & Eta.Rdm.Directory.ServiceFilterFlags.STATE) == Eta.Rdm.Directory.ServiceFilterFlags.STATE))
                        {

                            if (itemInfo == null)
                            {
                                itemInfo = m_ServerBaseImpl.ServerPool.GetItemInfo();

                                itemInfo.RequestMsg((IRequestMsg)directoryMsgEvent.Msg);
                                itemInfo.ClientSession = clientSession;

                                m_ServerBaseImpl.AddItemInfo(itemInfo);

                                ItemInfoList.Add(itemInfo);

                                if (m_apiAdminControl == false)
                                {
                                    var reqMsg = m_ServerBaseImpl.RequestMsg();

                                    reqMsg.Decode(directoryMsgEvent.Msg, directoryMsgEvent.ReactorChannel.MajorVersion,
                                        directoryMsgEvent.ReactorChannel.MinorVersion, null);

                                    m_ServerBaseImpl.OmmProviderEvent.clientHandle = clientSession.ClientHandle;
                                    m_ServerBaseImpl.OmmProviderEvent.closure = m_ServerBaseImpl.Closure;
                                    m_ServerBaseImpl.OmmProviderEvent.m_OmmProvider = m_ServerBaseImpl.Provider;
                                    m_ServerBaseImpl.OmmProviderEvent.handle = itemInfo.Handle;
                                    m_ServerBaseImpl.OmmProviderEvent.ReactorChannel = directoryMsgEvent.ReactorChannel;

                                    m_ServerBaseImpl.OmmProviderClient.OnAllMsg(reqMsg, m_ServerBaseImpl.OmmProviderEvent);
                                    m_ServerBaseImpl.OmmProviderClient.OnReqMsg(reqMsg, m_ServerBaseImpl.OmmProviderEvent);

                                    reqMsg.Clear();
                                }
                                else
                                {
                                    HandleDirectoryRequest(directoryMsgEvent.ReactorChannel, directoryRequest);
                                }
                            }
                            else
                            {
                                itemInfo.RequestMsg((IRequestMsg)directoryMsgEvent.Msg);

                                if (m_apiAdminControl == false)
                                {
                                    var reqMsg = m_ServerBaseImpl.RequestMsg();

                                    reqMsg.Decode(directoryMsgEvent.Msg, directoryMsgEvent.ReactorChannel.MajorVersion, directoryMsgEvent.ReactorChannel.MinorVersion, null);

                                    m_ServerBaseImpl.OmmProviderEvent.clientHandle = clientSession.ClientHandle;
                                    m_ServerBaseImpl.OmmProviderEvent.closure = m_ServerBaseImpl.Closure;
                                    m_ServerBaseImpl.OmmProviderEvent.m_OmmProvider = m_ServerBaseImpl.Provider;
                                    m_ServerBaseImpl.OmmProviderEvent.handle = itemInfo.Handle;
                                    m_ServerBaseImpl.OmmProviderEvent.ReactorChannel = directoryMsgEvent.ReactorChannel;

                                    m_ServerBaseImpl.OmmProviderClient.OnAllMsg(reqMsg, m_ServerBaseImpl.OmmProviderEvent);
                                    m_ServerBaseImpl.OmmProviderClient.OnReissue(reqMsg, m_ServerBaseImpl.OmmProviderEvent);

                                    reqMsg.Clear();
                                }
                                else
                                {
                                    HandleDirectoryRequest(directoryMsgEvent.ReactorChannel, directoryRequest);
                                }
                            }
                        }
                        else
                        {
                            temp.Append("Source directory request rejected - request message must minimally have SERVICE_INFO_FILTER and SERVICE_STATE_FILTER filters");

                            SendDirectoryReject(directoryMsgEvent.ReactorChannel, directoryMsgEvent.Msg.StreamId, StateCodes.USAGE_ERROR, temp.ToString());

                            if (m_ServerBaseImpl.GetLoggerClient().IsTraceEnabled)
                            {
                                temp.Append(ILoggerClient.CR).Append("stream Id ").Append(directoryMsgEvent.Msg.StreamId)
                                .Append(ILoggerClient.CR).Append("client handle ").Append(clientSession.ClientHandle);
                                m_ServerBaseImpl.GetLoggerClient().Trace(CLIENT_NAME, temp.ToString());
                            }

                            if (itemInfo != null)
                            {
                                if (m_apiAdminControl == false)
                                {
                                    NotifyOnClose(directoryMsgEvent.ReactorChannel, directoryMsgEvent.Msg, itemInfo);
                                }

                                ItemInfoList.Remove(itemInfo);
                                m_ServerBaseImpl.RemoveItemInfo(itemInfo, false);
                            }
                        }

                        break;
                    }
                case DirectoryMsgType.CONSUMER_STATUS:
                    {
                        if (m_ServerBaseImpl.GetLoggerClient().IsTraceEnabled)
                        {
                            temp.Append("Received directory consumer status message.")
                                .Append(ILoggerClient.CR).Append("Stream Id ").Append(directoryMsgEvent.Msg.StreamId)
                                .Append(ILoggerClient.CR).Append("Client handle ").Append(clientSession.ClientHandle)
                                .Append(ILoggerClient.CR).Append("Instance Name ").Append(m_ServerBaseImpl.InstanceName);

                            m_ServerBaseImpl.GetLoggerClient().Trace(CLIENT_NAME, temp.ToString());
                        }

                        itemInfo = clientSession.GetItemInfo(directoryMsgEvent.Msg.StreamId);

                        if (itemInfo != null)
                        {
                            GenericMsg genericMsg = m_ServerBaseImpl.GenericMsg();

                            genericMsg.Decode(directoryMsgEvent.Msg, directoryMsgEvent.ReactorChannel.MajorVersion, directoryMsgEvent.ReactorChannel.MinorVersion, null);

                            m_ServerBaseImpl.OmmProviderEvent.handle = clientSession.ClientHandle;
                            m_ServerBaseImpl.OmmProviderEvent.closure = m_ServerBaseImpl.Closure;
                            m_ServerBaseImpl.OmmProviderEvent.m_OmmProvider = m_ServerBaseImpl.Provider;
                            m_ServerBaseImpl.OmmProviderEvent.handle = itemInfo.Handle;
                            m_ServerBaseImpl.OmmProviderEvent.ReactorChannel = directoryMsgEvent.ReactorChannel;

                            m_ServerBaseImpl.OmmProviderClient.OnAllMsg(genericMsg, m_ServerBaseImpl.OmmProviderEvent);
                            m_ServerBaseImpl.OmmProviderClient.OnGenericMsg(genericMsg, m_ServerBaseImpl.OmmProviderEvent);

                            genericMsg.Clear();
                        }

                        break;
                    }
                case DirectoryMsgType.CLOSE:
                    {
                        if (m_ServerBaseImpl.GetLoggerClient().IsTraceEnabled)
                        {
                            temp.Append("Received directory close message.")
                                .Append(ILoggerClient.CR).Append("Stream Id ").Append(directoryMsgEvent.Msg.StreamId)
                                .Append(ILoggerClient.CR).Append("Client handle ").Append(clientSession.ClientHandle)
                                .Append(ILoggerClient.CR).Append("Instance Name ").Append(m_ServerBaseImpl.InstanceName);

                            m_ServerBaseImpl.GetLoggerClient().Trace(CLIENT_NAME, temp.ToString());
                        }

                        itemInfo = clientSession.GetItemInfo(directoryMsgEvent.Msg.StreamId);

                        if (itemInfo != null)
                        {
                            if (m_apiAdminControl == false)
                            {
                                m_ServerBaseImpl.OmmProviderEvent.ReactorChannel = directoryMsgEvent.ReactorChannel;
                                NotifyOnClose(directoryMsgEvent.ReactorChannel, directoryMsgEvent.Msg, itemInfo);
                            }

                            ItemInfoList.Remove(itemInfo);
                            m_ServerBaseImpl.RemoveItemInfo(itemInfo, false);
                        }

                        break;
                    }
                default:
                    temp.Append("Rejected unhandled directory message type ").Append(directoryMsg.DirectoryMsgType.ToString());

                    itemInfo = clientSession.GetItemInfo(directoryMsgEvent.Msg.StreamId);

                    if (itemInfo == null)
                    {
                        SendDirectoryReject(directoryMsgEvent.ReactorChannel, directoryMsgEvent.Msg.StreamId, StateCodes.USAGE_ERROR, temp.ToString());
                    }

                    if (m_ServerBaseImpl.GetLoggerClient().IsTraceEnabled)
                    {
                        temp.Append(ILoggerClient.CR).Append("Stream Id ").Append(directoryMsgEvent.Msg.StreamId)
                           .Append(ILoggerClient.CR).Append("Client handle ").Append(clientSession.ClientHandle)
                           .Append(ILoggerClient.CR).Append("Instance Name ").Append(m_ServerBaseImpl.InstanceName);

                        m_ServerBaseImpl.GetLoggerClient().Trace(CLIENT_NAME, temp.ToString());
                    }

                    break;
            }

            return ReactorCallbackReturnCode.SUCCESS;
        }

        void NotifyOnClose(ReactorChannel reactorChannel, Eta.Codec.IMsg msg, ItemInfo itemInfo)
        {
            IRequestMsg rsslReqMsg = m_ServerBaseImpl.GetRequestMsg();

            rsslReqMsg.ApplyNoRefresh();

            rsslReqMsg.StreamId = msg.StreamId;

            if (itemInfo.MsgKey.CheckHasName())
            {
                rsslReqMsg.MsgKey.ApplyHasName();
                rsslReqMsg.MsgKey.Name = itemInfo.MsgKey.Name;
            }

            if (itemInfo.MsgKey.CheckHasNameType())
            {
                rsslReqMsg.MsgKey.ApplyHasNameType();
                rsslReqMsg.MsgKey.NameType = itemInfo.MsgKey.NameType;
            }

            rsslReqMsg.DomainType = msg.DomainType;

            var reqMsg = m_ServerBaseImpl.RequestMsg();

            if (itemInfo.MsgKey.CheckHasServiceId())
            {
                rsslReqMsg.MsgKey.ApplyHasServiceId();
                rsslReqMsg.MsgKey.ServiceId = itemInfo.MsgKey.ServiceId;

                reqMsg.Decode(rsslReqMsg, reactorChannel.MajorVersion, reactorChannel.MinorVersion, null);

                m_ServerBaseImpl.GetDirectoryServiceStore().GetServiceNameById(itemInfo.MsgKey.ServiceId, out string? serviceName);

                if (serviceName != null)
                {
                    reqMsg.m_rsslMsg.MsgKey.Flags &= ~MsgKeyFlags.HAS_SERVICE_ID;
                    reqMsg.ServiceName(serviceName);
                    reqMsg.m_rsslMsg.MsgKey.Flags |= MsgKeyFlags.HAS_SERVICE_ID;
                }
            }
            else
            {
                reqMsg.Decode(rsslReqMsg, reactorChannel.MajorVersion, reactorChannel.MinorVersion, null);
            }

            int flags = reqMsg.m_rsslMsg.Flags;
            flags &= ~RequestMsgFlags.STREAMING;
            rsslReqMsg.Flags = flags;

            m_ServerBaseImpl.OmmProviderEvent.clientHandle = itemInfo.ClientSession!.ClientHandle;
            m_ServerBaseImpl.OmmProviderEvent.closure = m_ServerBaseImpl.Closure;
            m_ServerBaseImpl.OmmProviderEvent.m_OmmProvider = m_ServerBaseImpl.Provider;
            m_ServerBaseImpl.OmmProviderEvent.handle = itemInfo.Handle;

            m_ServerBaseImpl.OmmProviderClient.OnAllMsg(reqMsg, m_ServerBaseImpl.OmmProviderEvent);
            m_ServerBaseImpl.OmmProviderClient.OnClose(reqMsg, m_ServerBaseImpl.OmmProviderEvent);

            reqMsg.Clear();
        }

        void SendDirectoryReject(ReactorChannel reactorChannel, int streamId, int statusCode, string text)
        {
            m_directoryStatus.Clear();
            m_directoryStatus.StreamId = streamId;
            m_directoryStatus.HasState = true;
            m_directoryStatus.State.StreamState(StreamStates.CLOSED_RECOVER);
            m_directoryStatus.State.DataState(DataStates.SUSPECT);
            m_directoryStatus.State.Code(statusCode);
            Buffer buffer = new Buffer();
            buffer.Data(text);
            m_directoryStatus.State.Text(buffer);

            m_rsslSubmitOptions.Clear();

            ReactorReturnCode retCode = reactorChannel.Submit(m_directoryStatus, m_rsslSubmitOptions, out var errorInfo);

            if (retCode != ReactorReturnCode.SUCCESS)
            {
                if (m_ServerBaseImpl.GetLoggerClient().IsErrorEnabled)
                {
                    using var lockScope = m_ServerBaseImpl.GetUserLocker().EnterLockScope();
                    StringBuilder temp = m_ServerBaseImpl.GetStrBuilder();
                    Eta.Transports.Error? error = errorInfo?.Error;
                    ClientSession clientSession = (ClientSession)reactorChannel.UserSpecObj!;

                    temp.Append("Internal error: ReactorChannel.Submit() failed in DirectoryHandler.SendDirectoryReject().")
                        .Append(ILoggerClient.CR).Append("Client handle ").Append(clientSession.ClientHandle)
                        .Append(ILoggerClient.CR).Append("Instance Name ").Append(m_ServerBaseImpl.InstanceName)
                        .Append(ILoggerClient.CR).Append("Error Id ").Append(error?.ErrorId)
                        .Append(ILoggerClient.CR).Append("Internal sysError ").Append(error?.SysError)
                        .Append(ILoggerClient.CR).Append("Error Location ").Append(errorInfo?.Location)
                        .Append(ILoggerClient.CR).Append("Error Text ").Append(error?.Text);

                    m_ServerBaseImpl.GetLoggerClient().Error(CLIENT_NAME, temp.ToString());
                }
            }
        }

        void HandleDirectoryRequest(ReactorChannel reactorChannel, DirectoryRequest directoryRequest)
        {
            m_directoryRefresh.Clear();
            m_directoryRefresh.State.StreamState(StreamStates.OPEN);
            m_directoryRefresh.State.DataState(DataStates.OK);
            m_directoryRefresh.State.Code(StateCodes.NONE);
            m_directoryRefresh.State.Text(m_statusText);

            long filters = m_ServerBaseImpl.GetDirectoryServiceStore().EncodeDirectoryMsg(
                m_ServerBaseImpl.GetDirectoryServiceStore().GetDirectoryCache().DirectoryRefresh.ServiceList, 
                m_directoryRefresh.ServiceList, 
                directoryRequest.Filter, 
                true, 
                directoryRequest.HasServiceId, 
                directoryRequest.ServiceId);

            m_directoryRefresh.Filter = filters;
            m_directoryRefresh.ClearCache = true;
            m_directoryRefresh.StreamId = directoryRequest.StreamId;
            m_directoryRefresh.Solicited = true;

            m_rsslSubmitOptions.Clear();

            ReactorReturnCode retCode = reactorChannel.Submit(m_directoryRefresh, m_rsslSubmitOptions, out var errorInfo);

            if (retCode != ReactorReturnCode.SUCCESS)
            {
                if (m_ServerBaseImpl.GetLoggerClient().IsErrorEnabled)
                {
                    using var lockScope = m_ServerBaseImpl.GetUserLocker().EnterLockScope();
                    StringBuilder temp = m_ServerBaseImpl.GetStrBuilder();
                    Eta.Transports.Error? error = errorInfo?.Error;

                    ClientSession clientSession = (ClientSession)reactorChannel.UserSpecObj!;

                    temp.Append("Internal error: rsslChannel.Submit() failed in DirectoryHandler.HandleDirectoryRequest().")
                        .Append(ILoggerClient.CR).Append("Client handle ").Append(clientSession.ClientHandle)
                        .Append(ILoggerClient.CR).Append("Instance Name ").Append(m_ServerBaseImpl.InstanceName)
                        .Append(ILoggerClient.CR).Append("Error Id ").Append(error?.ErrorId)
                        .Append(ILoggerClient.CR).Append("Internal sysError ").Append(error?.SysError)
                        .Append(ILoggerClient.CR).Append("Error Location ").Append(errorInfo?.Location)
                        .Append(ILoggerClient.CR).Append("Error Text ").Append(error?.Text);

                    m_ServerBaseImpl.GetLoggerClient().Error(CLIENT_NAME, temp.ToString());

                    temp.Length = 0;
                }
            }

            m_ServerBaseImpl.GetDirectoryServiceStore().ReturnServiceToPool(m_directoryRefresh.ServiceList);
        }
    }
}
